packages/server/src/utils/buildChatflow.ts

```diff
@@ -187,45 +187,21 @@ export const utilBuildChatflow = async (req: Request, socketIO?: Server, isInter
             /*** Get Ending Node with Directed Graph  ***/
             const { graph, nodeDependencies } = constructGraphs(nodes, edges)
             const directedGraph = graph
-            const endingNodeIds = getEndingNodes(nodeDependencies, directedGraph)
-            if (!endingNodeIds.length) {
-                return {
-                    executionError: true,
-                    status: 500,
-                    msg: `Ending nodes not found`
-                }
-            }
 
-            const endingNodes = nodes.filter((nd) => endingNodeIds.includes(nd.id))
+            const [ok, endingNodesOrError] = getEndingNodes(nodeDependencies, directedGraph, nodes)
+            if (!ok) {
+                return endingNodesOrError
+            }
+            const endingNodes = endingNodesOrError
 
-            let isEndingNodeExists = endingNodes.find((node) => node.data?.outputs?.output === 'EndingNode')
+            let isEndingNodeExists = endingNodes.some((node) => node.data?.outputs?.output === 'EndingNode')
 
             for (const endingNode of endingNodes) {
                 const endingNodeData = endingNode.data
-                if (!endingNodeData) {
-                    return {
-                        executionError: true,
-                        status: 500,
-                        msg: `Ending node ${endingNode.id} data not found`
-                    }
-                }
 
                 const isEndingNode = endingNodeData?.outputs?.output === 'EndingNode'
 
                 if (!isEndingNode) {
-                    if (
-                        endingNodeData &&
-                        endingNodeData.category !== 'Chains' &&
-                        endingNodeData.category !== 'Agents' &&
-                        endingNodeData.category !== 'Engine'
-                    ) {
-                        return {
-                            executionError: true,
-                            status: 500,
-                            msg: `Ending node must be either a Chain or Agent`
-                        }
-                    }
-
                     if (
                         endingNodeData.outputs &&
                         Object.keys(endingNodeData.outputs).length &&
```

Comment:
@HenryHengZJ I'm not sure if this needs to be moved to "getEndingNodes" as well

Reply:
@mokeyish we can leave it here, because its not required when checking if chatflow is streamable